core/src/connection/pool.rs

```diff
@@ -763,49 +763,39 @@ where
                         ),
                     };
 
-                    enum Error {
-                        ConnectionLimit(ConnectionLimit),
-                        InvalidPeerId,
-                    }
-
-                    impl<TransportError> From<Error> for PendingConnectionError<TransportError> {
-                        fn from(error: Error) -> Self {
-                            match error {
-                                Error::ConnectionLimit(limit) => {
-                                    PendingConnectionError::ConnectionLimit(limit)
-                                }
-                                Error::InvalidPeerId => PendingConnectionError::InvalidPeerId,
-                            }
-                        }
-                    }
-
-                    let error = self
+                    let error: Result<(), PendingInboundConnectionError<_>> = self
                         .counters
                         // Check general established connection limit.
                         .check_max_established(&endpoint)
-                        .map_err(Error::ConnectionLimit)
+                        .map_err(PendingConnectionError::ConnectionLimit)
                         // Check per-peer established connection limit.
                         .and_then(|()| {
                             self.counters
                                 .check_max_established_per_peer(num_peer_established(
                                     &self.established,
                                     peer_id,
                                 ))
-                                .map_err(Error::ConnectionLimit)
+                                .map_err(PendingConnectionError::ConnectionLimit)
                         })
                         // Check expected peer id matches.
                         .and_then(|()| {
                             if let Some(peer) = expected_peer_id {
                                 if peer != peer_id {
-                                    return Err(Error::InvalidPeerId);
+                                    return Err(PendingConnectionError::InvalidPeerId {
+                                        obtained: peer_id,
+                                        address: endpoint.get_remote_address().clone(),
+                                    });
                                 }
                             }
                             Ok(())
                         })
                         // Check peer is not local peer.
                         .and_then(|()| {
                             if self.local_id == peer_id {
-                                Err(Error::InvalidPeerId)
+                                Err(PendingConnectionError::InvalidPeerId {
+                                    obtained: peer_id,
```

Comment:
How about renaming `peer_id` to `obtained_peer_id`? That would be consistent with `expected_peer_id` and would be less confusing.

Reply:
yes, will do that once I get a minute — crunch time right now